torch/csrc/jit/profiling_record.cpp

```diff
@@ -0,0 +1,79 @@
+#include <torch/csrc/jit/profiling_record.h>
+
+namespace torch {
+namespace jit {
+
+ProfilingRecord::ProfilingRecord(const std::shared_ptr<Graph>& g)
+    : profiled_graph_(g) {}
+
+void ProfilingRecord::instrumentBlock(Block* block) {
+  auto graph = block->owningGraph();
+
+  // iterating backwards allows us to easily insert profile nodes
+  // without affecting an iterator
+  for (auto it = block->nodes().rend(); it != block->nodes().rbegin(); --it) {
+    auto n = *it;
+    for (auto o : n->outputs()) {
+      if (!o->type()->isSubclass(TypeKind::TensorType)) {
+        continue;
+      }
+      auto pn = graph->create(prim::profile, 0);
+
+      std::function<void(Stack&)> shape_profiler = [this, o](Stack& stack) {
+        IValue t;
+        pop(stack, t);
+        if (t.isTensor()) {
+          auto pttp = ProfiledTensorType::create(t.toTensor());
+          std::lock_guard<std::mutex> lock(this->mutex_);
+          if (o->type()->isSubclass(TypeKind::ProfiledTensorType)) {
+            auto type = o->type()->cast<ProfiledTensorType>();
+            o->setType(type->merge(pttp));
+          } else {
+            o->setType(pttp);
+          }
+        }
+      };
+
+      pn->i_(attr::data, addCallback(shape_profiler));
+      pn->addInput(o);
+      pn->insertAfter(n);
+    }
+
+    for (auto b : n->blocks()) {
+      instrumentBlock(b);
+    }
+  }
+}
+
+std::unique_ptr<ProfilingRecord> ProfilingRecord::instrumentGraph(
+    const std::shared_ptr<Graph>& graph) {
+  auto new_g = graph->copy();
+  auto pr = std::unique_ptr<ProfilingRecord>(new ProfilingRecord(new_g));
+  auto raw_pr = pr.get();
+
+  pr->instrumentBlock(new_g->block());
+  std::function<void(Stack&)> counter = [raw_pr](Stack&) {
+    std::lock_guard<std::mutex> lock(raw_pr->mutex_);
+    raw_pr->profiling_count_--;
+  };
+
+  auto pop = new_g->create(prim::profile, 0);
```

Comment:
note how the details (attribute name, inputs, etc.) of prim::profile get repeated here.